src/utils.jl

```diff
@@ -26,113 +26,119 @@ end
     return output
 end
 
-Base.float(q::AbstractQuantity) = new_quantity(typeof(q), float(ustrip(q)), dimension(q))
-Base.convert(::Type{T}, q::AbstractQuantity) where {T<:Real} =
+Base.float(q::AbstractUnionQuantity) = new_quantity(typeof(q), float(ustrip(q)), dimension(q))
+Base.convert(::Type{T}, q::AbstractUnionQuantity) where {T<:Number} =
     let
         @assert iszero(dimension(q)) "$(typeof(q)): $(q) has dimensions! Use `ustrip` instead."
         return convert(T, ustrip(q))
     end
-Base.promote_rule(::Type{Dimensions{R1}}, ::Type{Dimensions{R2}}) where {R1,R2} = Dimensions{promote_type(R1,R2)}
-Base.promote_rule(::Type{Q1}, ::Type{Q2}) where {T1,T2,D1,D2,Q1<:Quantity{T1,D1},Q2<:Quantity{T2,D2}} = Quantity{promote_type(T1,T2),promote_type(D1,D2)}
+function Base.promote_rule(::Type{Dimensions{R1}}, ::Type{Dimensions{R2}}) where {R1,R2}
+    return Dimensions{promote_type(R1,R2)}
+end
+function Base.promote_rule(::Type{<:GenericQuantity{T1,D1}}, ::Type{<:GenericQuantity{T2,D2}}) where {T1,T2,D1,D2}
+    return GenericQuantity{promote_type(T1,T2),promote_type(D1,D2)}
+end
+function Base.promote_rule(::Type{<:Quantity{T1,D1}}, ::Type{<:GenericQuantity{T2,D2}}) where {T1,T2,D1,D2}
+    return GenericQuantity{promote_type(T1,T2),promote_type(D1,D2)}
+end
+function Base.promote_rule(::Type{<:Quantity{T1,D1}}, ::Type{<:Quantity{T2,D2}}) where {T1,T2,D1,D2}
+    return Quantity{promote_type(T1,T2),promote_type(D1,D2)}
+end
+function Base.promote_rule(::Type{T1}, ::Type{<:Quantity{T2,D2}}) where {T1<:Number,T2,D2}
```

Comment:
If this rule is needed for `Quantity`, should there also be an analogous promotion rule for `GenericQuantity`? I think that promotion with generic quantities might be one of the most important things to add a few tests for in this PR.

Reply:
Good point.

Reply:
Deleted this promotion rule as I realized it was way too powerful, and would basically turn everything it touches into Quantities. Instead should just promote to a union type or `Any` (already provided by `Base`), or have the user write a promotion method for their type.

Reply:
Actually, wait, this is needed, otherwise the code

```julia
[0.5, 0.5u"km"]
```

will return an error (since it will try to convert to `Float64`)... Hmm...

Reply:
Hm, I'm not sure I follow -- what was wrong with turning numbers into quantities? Then `[0.5, 0.5u"km"]` would for example be concretly typed, which seems to fit the spirit of removing type stability footguns

Reply:
My concern is that if you had this behavior in other abstract number packages, there would an ambiguity about which type takes priority. For example, say you had

```julia
[0.5 ± 0.1, 0.5u"km"]
```

would this get promoted to a `Vector{Measurement{Quantity{Float64,D}}}` or a `Vector{Quantity{Measurement{Float64},D}}`?

With this PR, it simply gets promoted to a `Vector{Number}` which I think is more conservative. Thoughts?

Reply:
What if we did the more aggressive thing for types `T` satisfying `base_number_type(T) = T`?

Reply:
But I'd also be happy punting this and making an issue for more careful thought about promotion and its relation to array types. The behaviour here is indeed slightly better than `main`, so I don't think anything must be addressed in this PR in any case. (Perhaps too idealistically, it would be nice to even figure out how to make something concrete in cases with measurements etc. The principle being that if `[0.5 ± 0.1, 0.2]` makes something concrete, then replacing `0.2` with a quantity should not hurt type stability.)

Reply:
> What if we did the more aggressive thing for types `T` satisfying `base_number_type(T) = T`?

That's a good idea. Although what would happen if you have `[1, 1.0, 1.0u"km"]`? Would it first try to promote `[1, 1.0]` to `[1.0, 1.0]` and then move all to quantities? And would it differ from `[1, 1.0u"km", 1.0]`?

Reply:
Punting it sounds good!